encoding/encoding.go

```diff
@@ -12,38 +17,189 @@ type Decodable interface {
 
 // Encoder represents types that can encode values.
 type Encoder interface {
-	// EncodeObject encodes an object.
-	EncodeObject(obj Encodable) error
-	// IntoBytes returns the encoded bytes.
-	IntoBytes() []byte
+	// EncodeUint8 encodes a uint8.
+	EncodeUint8(obj uint8) error
+	// EncodeUint16 encodes a uint16.
+	EncodeUint16(obj uint16) error
+	// EncodeUint32 encodes a uint32.
+	EncodeUint32(obj uint32) error
+	// EncodeUint64 encodes a uint64.
+	EncodeUint64(obj uint64) error
+	// EncodeInt8 encodes a int8.
+	EncodeInt8(obj int8) error
+	// EncodeInt16 encodes a int16.
+	EncodeInt16(obj int16) error
+	// EncodeInt32 encodes a int32.
+	EncodeInt32(obj int32) error
+	// EncodeInt64 encodes a int64.
+	EncodeInt64(obj int64) error
+	// EncodeBool encodes a bool.
+	EncodeBool(obj bool) error
+	// EncodeString encodes a string.
+	EncodeString(obj string) error
+	// EncodeArray encodes an array.
+	EncodeArray(obj interface{}) error
+	// EncodeMap encodes a map.
+	EncodeMap(obj interface{}) error
+	// EncodeStruct encodes a struct.
+	EncodeStruct(obj interface{}) error
+	// EncodeTag encodes a value with a name.
+	EncodeTag(name interface{}, value interface{}) error
+	// Bytes returns the encoded bytes.
+	Bytes() []byte
 }
 
 // Decoder represents types that can decode values.
 type Decoder interface {
-	// SetBytes sets the initializer internal bytes to match the input.
-	SetBytes([]byte)
-	// DecodeObject decodes an object.
-	DecodeObject(obj Decodable) error
+	// DecodeValue encodes a primitive value.
+	DecodeValue(obj interface{}) error
+	// DecodeArray encodes an array.
+	DecodeArray(obj interface{}) error
+	// DecodeMap encodes a map.
+	DecodeMap(obj interface{}) error
+	// DecodeStruct encodes a struct.
+	DecodeStruct(obj interface{}) error
 }
 
 type defaultEncoder = IpldCborEncoder
 type defaultDecoder = IpldCborDecoder
 
-// Encode encodes an encodable type, returning a byte array.
-func Encode(obj Encodable) ([]byte, error) {
+// Encode encodes an object, returning a byte array.
+func Encode(obj interface{}) ([]byte, error) {
 	var encoder Encoder = &defaultEncoder{}
+	return encode(obj, reflect.ValueOf(obj), encoder)
+}
+
+// EncodeWith encodes an object using the encoder provided returning a byte array.
+func EncodeWith(obj interface{}, encoder Encoder) ([]byte, error) {
+	return encode(obj, reflect.ValueOf(obj), encoder)
+}
+
+func encode(obj interface{}, v reflect.Value, encoder Encoder) ([]byte, error) {
+	var err error
+
+	// if `Encodable`, we are done
+	if encodable, ok := obj.(Encodable); ok {
+		if err = encodable.Encode(encoder); err != nil {
+			return nil, err
+		}
+		return encoder.Bytes(), nil
+	}
+
+	switch v.Kind() {
+	case reflect.Uint:
+		// Review: is the size of `uint` fixed or depends on the architecture?
```

Comment:
Both int and uint are platform dependent.  You can read with [unsafe](https://golang.org/pkg/unsafe/#Sizeof)